working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1133,35 +1101,153 @@ if (var [int x, int y] = json) {
-That is less important here since the only other case is the else branch.
+    *All of these patterns are refutable and may fail to match. In a matching
+    context like a switch case, if a pattern fails to match, execution skips
+    over the case body to ensure that variables bound by the pattern can only
+    be used when the pattern matches. Declaration and assignment contexts have
+    no control flow, so they can only use patterns that will always match.*
+
+    *Logical-or patterns are refutable because there is no point in using one
+    with an irrefutable left operand. We could make null-check patterns
+    irrefutable if `V` is assignable to its static type, but whenever that is
+    true the pattern does nothing useful since its only behavior is a type
+    test.*
+
+    *The remaining patterns are allowed syntactically to appear in a refutable
+    context. Patterns that do type tests like variables and lists produce a
+    compile-time error when used in an irrefutable context if the matched value
+    isn't assignable to their required type. This error is specified under type
+    checking.*
+
+*   It is a compile-time error if a variable pattern in a declaration context is
+    marked `var` or `final`. *A pattern declaration statement is already
+    preceded by `var` or `final`, so allowing those on the variable patterns
+    inside would lead to unnecessary or confusing code like:*
+
+    ```dart
+    // Disallowed:
+    var [var x] = [1];
+    final [var y] = [2];
+    ```
+
+    *To declare variables in a declaration context, use a simple identifer:*
+
+    ```dart
+    // OK:
+    var [x] = [1];
+    final [y] = [2];
+    ```
+
+*   It is a compile-time error if a variable pattern in an assignment context is
+    marked `var`, `final`, or with a type annotation. *Patterns in assignments
+    can only assign to existing variables, not declare new ones.*
+
+    ```dart
+    var a = 1;
+    var b = 2;
+
+    // Disallowed:
+    (var a, int b) = (3, 4);
+
+    // OK:
+    (a, b) = (3, 4);
+    ```
+
+*   A simple identifier in a matching context is treated as a named constant
+    pattern unless its name is `_`. *A bare identifier is ambiguous and could
+    be either a named constant or a variable pattern without any `var`, `final`,
+    or type annotation marker. We prefer the constant interpretation for
+    backwards compatibility and to make variable declarations more explicit in
+    cases. To declare variables in a matching context, use `var`, `final`, or a
+    type before the name.*
+
+    *There is no ambiguity with bare identifiers in irrefutable contexts since
+    constant patterns are disallowed there.*
+
+    ```dart
+    const c = 1;
+    switch (2) {
+      case c: print('match $c');
+      default: print('no match');
+    }
+    ```
+
+    *This program prints "no match" and not "match 2".*
+
+*   A simple identifier in a matching context named `_` is treated as a wildcard
```

Comment:
Technically this is a breaking change, since in Dart today we can do:

```dart
main() {
  const _ = 1;
  switch (2) {
    case _:
      print('Matched _');
      break;
    default:
      print('Did not match _');
      break;
  }
}
```

(and this program prints `Did not match _`).

I think this is fine, because the chances of the breakage affecting a real user are so low, but we should probably mention in this doc that it's a breakage just for completeness.

Reply:
Good point. I went ahead and consolidated all of the breaking stuff into a new section.